diff --git a/src/Control/Monad/Free/Class.hs b/src/Control/Monad/Free/Class.hs
index ba1677d..ee60b55 100644
--- a/src/Control/Monad/Free/Class.hs
+++ b/src/Control/Monad/Free/Class.hs
@@ -149,6 +149,7 @@ instance (Functor f, MonadFree f m) => MonadFree f (ListT m) where
 #endif

 -- | A version of lift that can be used with just a Functor for f.
+{-# INLINE liftF #-}
 liftF :: (Functor f, MonadFree f m) => f a -> m a
 liftF = wrap . fmap return

@@ -156,5 +157,6 @@ liftF = wrap . fmap return
 --
 -- /Note:/ that this is the default implementation for 'wrap' for
 -- @MonadFree f (t m)@.
+{-# INLINE wrapT #-}
 wrapT :: (Functor f, MonadFree f m, MonadTrans t, Monad (t m)) => f (t m a) -> t m a
 wrapT = join . lift . liftF
