diff --git a/src/Control/State/Transition/Extended.hs b/src/Control/State/Transition/Extended.hs
index c21d242..8fa10a8 100644
--- a/src/Control/State/Transition/Extended.hs
+++ b/src/Control/State/Transition/Extended.hs
@@ -439,6 +439,7 @@ labeledPredE lbls cond orElse = labeled lbls (cond ?!: orElse)
 labeled :: NonEmpty Label -> Rule sts ctx () -> Rule sts ctx ()
 labeled lbls subrule = liftF $ Label lbls subrule ()

+{-# INLINE trans #-}
 trans ::
   Embed sub super => RuleContext rtype sub -> Rule super rtype (State sub)
 trans ctx = wrap $ SubTrans ctx pure
@@ -449,6 +450,7 @@ ifFailureFree x y = liftF (IfFailureFree x y)
 whenFailureFree :: Rule sts rtype () -> Rule sts rtype ()
 whenFailureFree action = ifFailureFree action (pure ())

+{-# INLINE liftSTS #-}
 liftSTS ::
   STS sts =>
   (BaseM sts) a ->
@@ -456,6 +458,7 @@ liftSTS ::
 liftSTS f = wrap $ Lift f pure

 -- | Get the judgment context
+{-# INLINE judgmentContext #-}
 judgmentContext :: Rule sts rtype (RuleContext rtype sts)
 judgmentContext = wrap $ GetCtx pure

